OCPEDGE-1310: feat: add arbiter node support to installer

eggfoobar opened the pull request:
Added initial support for the arbiter node installation path for baremetal. 

The arbiter node is similar to the master node and is part of the control plane but will be used for smaller hardware sizes. Currently this is a TechPreview feature and only supported on baremetal.

For more details see EP: https://github.com/openshift/enhancements/pull/1674

zaneb commented on pkg/asset/agent/installconfig.go:
```diff
@@ -215,9 +215,15 @@ func (a *OptionalInstallConfig) validateControlPlaneConfiguration(installConfig
+			// Make sure users get an accurate error message while feature is in TechPreview
+			// TODO: Remove once HighlyAvailableArbiterMode moves out of TechPreviewNoUpgrade and just use `"between 1 and 5"`
+			supportedControlPlaneRange := "to 5, 4, 3, or 1"
+			if installConfig.FeatureSet == configv1.TechPreviewNoUpgrade {
```
Note that ABI doesn't actually support FeatureSet yet - it fails with an error if you specify TechPreviewNoUpgrade because it doesn't have a mechanism to pass that on to assisted-service.
https://issues.redhat.com/browse/AGENT-554

This is probably the right thing to do though, as it means we will get the right behaviour if somebody implements that.

eggfoobar replied:
Other than having the capacity to work on it, are there technical reasons this would be a problem to implement? If it's wiring things up, I might have some cycles to push it through if we want this for ABI as TechPreview.

zaneb replied:
It's not technically difficult, it just hasn't reached the top of the priority list. I'd be more than happy for you to pick up that ticket. Let me or @rwsu know if you need any help.

eggfoobar replied:
Sounds good, I'll keep you posted, thanks!

zaneb commented on pkg/types/machinepools.go:
```diff
@@ -20,6 +20,8 @@ const (
 	// MachinePoolControlPlaneRoleName name associated with the control plane machinepool.
 	MachinePoolControlPlaneRoleName = "master"
+	// MachinePoolArbiterRoleName name associated with the control plane machinepool for smaller sized limited nodes.
+	MachinePoolArbiterRoleName = "arbiter"
```
Need to update the comments on ll. 56-57.

eggfoobar replied:
Thanks! Fixed

eggfoobar commented on pkg/infrastructure/aws/clusterapi/aws.go:
```diff
@@ -100,6 +100,7 @@ func (p Provider) Ignition(ctx context.Context, in clusterapi.IgnitionInput) ([]
 		clusterapi.IgnitionSecret(editedBootstrapIgn, in.InfraID, "bootstrap"),
 		clusterapi.IgnitionSecret(editedMasterIgn, in.InfraID, "master"),
 	}
+
```
Oops, hold out from the original AWS implementation. Removed

eggfoobar commented on pkg/asset/machines/baremetal/hosts.go:
```diff
@@ -172,3 +182,75 @@ func Hosts(config *types.InstallConfig, machines []machineapi.Machine, userDataS
+		return nil, nil
+	}
+
+	numRequiredControlPlaneMachines := len(machines)
```
Ah yup, copy and paste error. Kept going back and forth on whether to add the arbiter logic to the `Hosts` function instead of it's own, fixed.

eggfoobar commented on pkg/types/validation/installconfig.go:
```diff
@@ -668,6 +675,18 @@ func validateControlPlane(platform *types.Platform, pool *types.MachinePool, fld
+	}
+	allErrs = append(allErrs, ValidateMachinePool(platform, pool, fldPath)...)
+	return allErrs
+}
```
Good catch, fixed.

eggfoobar commented on pkg/types/baremetal/validation/platform.go:
```diff
@@ -333,6 +341,10 @@ func validateHostsCount(hosts []*baremetal.Host, installConfig *types.InstallCon
 		return fmt.Errorf("not enough hosts found (%v) to support all the configured ControlPlane replicas (%v)", numMasters, numRequiredMasters)
 	}
 
+	if numArbiters < numRequiredArbiters {
```
Oh I see, okay, updated if statement to include count for excess hosts.

eggfoobar commented on pkg/asset/machines/baremetal/hosts.go:
```diff
@@ -172,3 +182,75 @@ func Hosts(config *types.InstallConfig, machines []machineapi.Machine, userDataS
+			}
+
+			newHost.ObjectMeta.Labels = map[string]string{
+				"installer.openshift.io/role": "control-plane",
```
Yeah, we could update that service template to include an arbiter env count as well. Also to bring to your attention, in my mind the arbiter was technically part of the control plane due to etcd, so [this logic](https://github.com/openshift/installer/pull/9159/files#diff-2a27cf9f710e504a0f8b3710fa10d24e403ff92e781cc8acbb425571427cc51bR319-R322) adds that for the rendering of that shell script env, if we want to remove any ambiguity, I can update the script to wait for arbiters as well and we can add a more specific label for arbiters, wdyt?

zaneb replied:
> in my mind the arbiter was technically part of the control plane due to etcd

yeah, let's just keep it how you have it

eggfoobar commented on pkg/asset/machines/arbiter.go:
```diff
@@ -0,0 +1,421 @@
+}
+
+// IPAddresses returns IPAddress manifest structures.
+func (m *Arbiter) IPAddresses() ([]ipamv1.IPAddress, error) {
```
Honestly I don't know, I noticed it wasn't being used in the master machine either and thought it was needed by a dependency. If not, seems safe to chuck from what I can tell in this repo.

zaneb replied:
When it was [added](https://github.com/openshift/installer/commit/af931034328b603f29bf11b85d15b93db438aee1) it was used in tfvars.go (for vSphere only). Doesn't appear to be used any more since Terraform was replaced by CAPI providers. I think it is safe to delete.

eggfoobar commented on pkg/asset/machines/arbiter.go:
```diff
@@ -0,0 +1,421 @@
+	dependencies.Get(clusterID, installConfig, rhcosImage, mign)
+
+	// #nosec G101
+	arbiterUserDataSecretName := "arbiter-user-data" // nolint:ineffassign
```
So this is technically a default value, it's being used [further down](https://github.com/openshift/installer/pull/9159/files#diff-5faf02b1dde3e563bf9a859be65a666b16b49fb92281cc8fd79d16426dbb8fd9R179), in practice there is only one branch for the switch statement currently, so the linter would trigger since it's always being re-assigned, figured it would be best to leave the existing pattern that follows the other machines and quiet the linter, but I can remove the switch if that's preferable to quieting the linter.

eggfoobar replied:
On second thought, decided to remove the switch statement here and keep things straightforward.

eggfoobar commented on pkg/asset/machines/arbiter.go:
```diff
@@ -0,0 +1,421 @@
+
+		// Use managed user data secret, since we always have up to date images
+		// available in the cluster
+		arbiterUserDataSecretName = "arbiter-user-data-managed" // #nosec G101 //nolint:ineffassign
```
Oops, yup, removed.

eggfoobar commented on pkg/asset/machines/baremetal/hosts.go:
```diff
@@ -172,3 +182,75 @@ func Hosts(config *types.InstallConfig, machines []machineapi.Machine, userDataS
+	numMachines := 0
+	for _, host := range config.Platform.BareMetal.Hosts {
+		// We skip creating other host files, the master machine generation takes care of those
+		if !host.IsArbiter() {
```
So this gets tricky here, if I understand the flow correctly. When the user does not specify roles, the masters are essentially taking from the first `n` number of hosts to satisfy the master requirement, then the remainder are workers. Would it be correct to assume that it would be safe to grab the remaining hosts after the master nodes for the arbiter if no role strings are present?

eggfoobar replied:
In trying to approach this problem it gets a little messy to infer the intent during the generation phase. I was wondering if we can instantiate a role during config ingestion and validation part, this would allow us to set the Role if none is provided based off of the number of requested replicas, with order of importance going master -> arbiter -> worker(or empty). Since the field is optional, the problem becomes more messy when you have to solve for some Hosts having roles but not others. If we validate and instantiate during ingestion time we can solve that there when we have the total count desired and the host list, so by the time we get to the generation phase we can confidently use the Role field. What do you think am I over simplifying the issue?

eggfoobar replied:
For ex:

```go
	for i, h := range hosts {
		if h.IsMaster() {
			numMasters++
		} else if h.IsArbiter() {
			numArbiters++
		} else if h.IsWorker() {
			numWorkers++
		} else {
			logrus.Warn(fmt.Sprintf("Host %s hasn't any role configured", h.Name))
			if numMasters < numRequiredMasters {
				numMasters++
				hosts[i].Role = "master"
			} else if numArbiters < numRequiredArbiters {
				numArbiters++
				hosts[i].Role = "arbiter"
			} else if numWorkers < numRequiredWorkers {
				numWorkers++
				hosts[i].Role = "worker"
			}
		}
	}
```

zaneb replied:
The tricky part about just e.g. setting the roles in the install-config explicitly in `SetPlatformDefaults()` is that the hosts roles from the install-config are used by ABI, but if the roles are not present then for ABI we want assisted-service to choose them dynamically at runtime rather than statically in the installer.

Note that the reason the code above can work in a single pass is that we [sort the hosts by role](https://github.com/openshift/installer/blob/main/pkg/types/baremetal/defaults/platform.go#L141-L143) first. (You need to [handle](https://github.com/openshift/installer/blob/main/pkg/types/baremetal/platform.go#L58) the arbiter role case as well.)

eggfoobar replied:
Okay, I think this should be better now, the logic isn't the cleanest, but hopefully it's all clear. I added a unit tests to cover the bases and hopefully catch most permutations.

eggfoobar commented on pkg/asset/agent/installconfig.go:
```diff
@@ -215,9 +215,15 @@ func (a *OptionalInstallConfig) validateControlPlaneConfiguration(installConfig
+			// Make sure users get an accurate error message while feature is in TechPreview
+			// TODO: Remove once HighlyAvailableArbiterMode moves out of TechPreviewNoUpgrade and just use `"between 1 and 5"`
+			supportedControlPlaneRange := "to 5, 4, 3, or 1"
+			if installConfig.FeatureSet == configv1.TechPreviewNoUpgrade {
```
Oh that's a great suggestion, a lot cleaner. Updated

eggfoobar commented on pkg/asset/ignition/machine/arbiter.go:
```diff
@@ -0,0 +1,87 @@
+	"os"
+
+	igntypes "github.com/coreos/ignition/v2/config/v3_2/types"
+	"github.com/pkg/errors"
```
Sure thing, cleaning up the error package

eggfoobar commented on pkg/asset/ignition/machine/arbiter_ignition_customizations.go:
```diff
@@ -0,0 +1,92 @@
+		return errors.Wrap(err, "failed to Marshal defaultPointerIgnition")
+	}
+	if string(savedPointerIgnitionJSON) != string(defaultPointerIgnitionJSON) {
+		logrus.Infof("Arbiter pointer ignition was modified. Saving contents to a machineconfig")
```
Unfortunately I don't have the best context for you on this one @patrickdillon. Since these configs are mirroring master deployment paths, this part of the code mimics what we do for master nodes. From what I could gather this was done to capture a user modified an MC scenario. 

https://github.com/openshift/installer/pull/4413
> With this new approach, we detect the case where a change has
been made to the pointer config by the user, and in that case
it is stored via a MachineConfig manifest, such that any
customizations are reflected in the MCO rendered config.

patrickdillon replied:
Thanks for clearing this up. I didn't realize the MCO does actually generate a stub ignition as well as the normal ignition (although it makes perfect sense).

eggfoobar commented:
/retest-required

suleymanakbas91 commented:
/retest-required

patrickdillon commented on pkg/asset/machines/baremetal/hosts.go:
```diff
@@ -172,3 +194,104 @@ func Hosts(config *types.InstallConfig, machines []machineapi.Machine, userDataS
+}
+
+// isArbiterEnabled Determine if arbiter is enabled based off of the install config.
+func isArbiterEnabled(config *types.InstallConfig) bool {
+	return config.Arbiter != nil &&
+		config.Arbiter.Replicas != nil &&
+		*config.Arbiter.Replicas > 0
+}
```
nit/suggestion: might be more appropriate to define this as a pointer receiver on install config (or config.Arbiter if a pointer receiver can safely handle a nil receiver--not sure) especially if we reuse this logic

eggfoobar replied:
Oh that's a great idea, I don't think I can add it to the Arbiter attribute since that's a MachinePool, but the install config makes a lot of sense. Updating

patrickdillon commented on pkg/types/validation/installconfig.go:
```diff
@@ -749,6 +756,27 @@ func validateControlPlane(platform *types.Platform, pool *types.MachinePool, fld
+	if arbiterPool.Name != types.MachinePoolArbiterRoleName {
+		allErrs = append(allErrs, field.NotSupported(fldPath.Child("name"), arbiterPool.Name, []string{types.MachinePoolArbiterRoleName}))
+	}
+	if arbiterPool.Replicas != nil && *arbiterPool.Replicas == 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("replicas"), arbiterPool.Replicas, "number of arbiter replicas must be positive"))
+	}
```
this looks duplicated below

eggfoobar replied:
Ah, good catch, yup, removing duplicate

patrickdillon reviewed (commented):
By and large this looks good to me. I did leave one comment to delete some duplicate code, and one nit which you should feel free to ignore. 

What is the current testing status for this, both your local testing and do we have any CI tests?

The code looks pretty good to me but I would like to get another reviewer as well. I asked the metal platform if somebody could take a look, but also happy if @zaneb signs off.

patrickdillon commented:
> What is the current testing status for this, both your local testing and do we have any CI tests?

Looks like we don't have any tech preview baremetal tests available (which would only provide a very limited signal anyway).

eggfoobar commented:
> > What is the current testing status for this, both your local testing and do we have any CI tests?
> 
> Looks like we don't have any tech preview baremetal tests available (which would only provide a very limited signal anyway).

Yeah, right now we have validated things locally and have some [PRs in place for e2e tests](https://github.com/openshift/origin/pull/29380/files#diff-09d0b993b8fa8f5c4936cf06a1138ea59c9099988c4d1844485c213d35eea2cd) in origin but currently no CI running due to dependencies on other PRs

zaneb commented on pkg/asset/machines/baremetal/hosts.go:
```diff
@@ -116,10 +116,31 @@ func Hosts(config *types.InstallConfig, machines []machineapi.Machine, userDataS
+		if config.IsArbiterEnabled() {
+			// If we know the host is an arbiter and the role isn't an empty value (i.e. explicitly defined by user)
+			// no further investigating is needed, skip creating resources
+			if host.IsArbiter() && host.Role != "" {
```
nit: the `&& host.Role != ""` part is redundant here (since `IsArbiter()` returns `host.Role == "arbiter"`)

eggfoobar replied:
Ah, yup, removing.

zaneb commented on pkg/asset/machines/baremetal/hosts.go:
```diff
@@ -172,3 +193,97 @@ func Hosts(config *types.InstallConfig, machines []machineapi.Machine, userDataS
+		// When it comes to hosts, the masters should have been generated first we skip those hosts
+		// and grab the hosts after to satisfy the number of required arbiters, and the host isn't
+		// explicitly set as an arbiter, this allows for mixed role types in the list.
+		if numMasters < numRequiredMasters && !host.IsArbiter() {
```
nit: `&& !host.IsArbiter()` is redundant here; we would have already continued on line 225 if `host.IsArbiter()` were true.

eggfoobar replied:
This unfortunately is needed, but it's not very clear as to the intent here. I updated the logic to be what we're really checking for here, which is the situation where the `host.Role != ""` to account for empty roles correctly. I updated and hopefully added a clearer comment.

zaneb reviewed (commented):
Baremetal stuff looks good
/approve

dtantsur reviewed (commented):
Looks right from the baremetal perspective

patrickdillon reviewed (commented):
Hm, I see now that arbiter.ign is always being generated... even if no arbiter nodes are being used. I think it would make sense (and be less confusing for users) to only generate the arbiter ignition if the machine pool is populated? WDYT?

eggfoobar commented:
> Hm, I see now that arbiter.ign is always being generated... even if no arbiter nodes are being used. I think it would make sense (and be less confusing for users) to only generate the arbiter ignition if the machine pool is populated? WDYT?

Yeah that's a good catch, updated and squashed the commits as well. Thanks @patrickdillon!

suleymanakbas91 commented:
/retest-required

eggfoobar commented:
/retest-required

suleymanakbas91 commented:
/retest-required

eggfoobar commented:
/retest-required

patrickdillon commented:
/lgtm

eggfoobar commented:
/restest-required

eggfoobar commented:
/retest-required

patrickdillon commented:
/lgtm

eggfoobar commented:
/retest-required

suleymanakbas91 commented:
/retest-required

eggfoobar commented:
/retest-required

eggfoobar commented:
/retest-required

suleymanakbas91 commented:
/retest-required

eggfoobar commented:
/retest-required

eggfoobar commented:
/retest-required

eggfoobar commented:
/retest-required

eggfoobar commented:
/retest-required

eggfoobar commented:
/retest-required

eggfoobar commented:
/retest-required

eggfoobar commented:
/retest-required
